Lesson16_Task01.py

```diff
@@ -0,0 +1,136 @@
+# as minimum  one staticmethod, two or more getters/setters
+# do not forget about block '''if __name__ == '__main__':'''
+"""
+
+import datetime
+
+
+class BankCard:
+    """
+    This class represents a bank card with various attributes and operations.
+    """
+    PIN_SIZE = 4  # Example of a class attribute
+
+    def __init__(self, card_num, cvv, pin, name, surname, end_date):
+        self.card_num = card_num
+        self._cvv = cvv  # Protected attribute
+        self._pin = pin  # Protected attribute
+        self.name = name
+        self.surname = surname
+        self.end_date = end_date
+        self.is_blocked = False
+        self.balance = 0.0
+
+    @staticmethod
+    def generate_expiry_date():
+        """
+        Generate an expiry date for the card.
+        """
+        today = datetime.date.today()
+        expiry_date = today.replace(year=today.year + 5)
+        return expiry_date
+
+    @property
+    def cvv(self):
+        return self._cvv
+
+    @property
+    def pin(self):
+        return self._pin
+
+    def block_card(self):
+        """
+        Block the card.
+        """
+        self.is_blocked = True
+
+    def deposit(self, amount):
+        """
+        Deposit money into the card's balance.
+        """
+        if not self.is_blocked and amount > 0:
+            self.balance += amount
+        return self.balance
+
+    def withdraw(self, amount):
+        """
+        Withdraw money from the card's balance.
+        """
+        if not self.is_blocked and 0 < amount <= self.balance:
+            self.balance -= amount
+        return self.balance
+
+    # Getters and Setters
+    def get_balance(self):
+        """
+        Get the current balance of the card.
+        """
+        return self.balance
+
+    def get_full_name(self):
+        """
+        Get the full name of the card holder.
+        """
+        return f"{self.name} {self.surname}"
+
+    def set_pin(self, new_pin):
+        """
+        Set a new PIN for the card.
+        """
+        if len(str(new_pin)) == self.PIN_SIZE:
+            self._pin = new_pin
+            return True
+        return False
+
+    @staticmethod
+    def validate_card_number(card_num):
+        """
+        Validate the card number.
+        """
+        return len(card_num) == 16 and card_num.isdigit()
+
+
+if __name__ == "__main__":
+    card_num = "1234567890123456"
+    cvv = "123"
+    pin = "4321"
+    name = "John"
+    surname = "Doe"
+    end_date = BankCard.generate_expiry_date()
+
+    card = BankCard(card_num, cvv, pin, name, surname, end_date)
+
+    print(f"Card Holder: {card.get_full_name()}")
+    print(f"Card Number: {card.card_num}")
+    print(f"Expiry Date: {card.end_date}")
+    print(f"Current Balance: ${card.get_balance()}")
+
+    if card.validate_card_number(card.card_num):
+        print("Card number is valid.")
+    else:
+        print("Invalid card number.")
+
+    try:
```

Comment:
1 - your code has no raise of BaseException - I mean deposit and withdraw operation. But here you are waiting for it.
Deposit and withdraw methods should be updated i think
2 - there is no sense to add several try - except blocks - you able to place all your code within one try-except block